### PR TITLE
nixos/stage-1-init: Set host id for ZFS

### PR DESCRIPTION
**Pull request by dasJ**
###### Description of changes

Probably useful for people who want to mount ZFS in stage 1

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by ElvishJerricco — APPROVED** [review]

File: nixos/modules/tasks/network-interfaces.nix [hunk 1]
```diff
@@ -1383,16 +1392,8 @@ in
...
-          hi="${cfg.hostId}"
-          ${if pkgs.stdenv.isBigEndian then ''
-            echo -ne "\x''${hi:0:2}\x''${hi:2:2}\x''${hi:4:2}\x''${hi:6:2}" > $out
-          '' else ''
-            echo -ne "\x''${hi:6:2}\x''${hi:4:2}\x''${hi:2:2}\x''${hi:0:2}" > $out
-          ''}
-        '';
-      };
+    environment.etc.hostid = mkIf (cfg.hostId != null) { source = hostidFile; };
+    boot.initrd.systemd.contents."/etc/hostid" = mkIf (cfg.hostId != null) { source = hostidFile; };
```
Comment (lovesegfault) [inline nixos/modules/tasks/network-interfaces.nix:1396]: What happens if `cfg.hostId != null`? I think you want `source = cfg.environment.etc.hostid`?

Comment (dasJ) [reply nixos/modules/tasks/network-interfaces.nix:1396]: wdym? Are you mistaking cfg and config?

Comment (lovesegfault) [reply nixos/modules/tasks/network-interfaces.nix:1396]: Yup, I am!

**Comment by Luflosi** [comment]
Why is this needed? There already is some code that looks like it is trying to do exactly the same thing:
https://github.com/NixOS/nixpkgs/blob/3f27c0e57f97e3e096b3aeecf958a5f2bbd9477c/nixos/modules/system/boot/stage-1-init.sh#L227-L229 and https://github.com/NixOS/nixpkgs/blob/3f27c0e57f97e3e096b3aeecf958a5f2bbd9477c/nixos/modules/system/boot/stage-1.nix#L337-L345
Can this code be removed now?

**Comment by lovesegfault** [comment]
@Luflosi This is for the new systemd-based stage-1 boot, which is an alternative to the scripts-based one we currently have.

**Comment by Luflosi** [comment]
Ok, so as long as we have both stage-1 boot alternatives, we need both pieces of code (the one added in this PR and the one I linked above)?

**Comment by lovesegfault** [comment]
> Ok, so as long as we have both stage-1 boot alternatives, we need both pieces of code (the one added in this PR and the one I linked above)?

We might be able to deduplicate the `setHostId` method, but it'll still need to be threaded through to both stage-1 boots.

It seems like the duplication of `setHostId` predates this though, I wonder why.

**Comment by Luflosi** [comment]
Thanks for the explanation!

**Pull request merged**